clippy_lints/src/types/redundant_allocation.rs

```diff
@@ -25,19 +25,40 @@ pub(super) fn check(cx: &LateContext<'_>, hir_ty: &hir::Ty<'_>, qpath: &QPath<'_
         }
     }
 
-    if cx.tcx.is_diagnostic_item(sym::Rc, def_id) {
-        if let Some(ty) = is_ty_param_diagnostic_item(cx, qpath, sym::Rc) {
-            let mut applicability = Applicability::MachineApplicable;
-            span_lint_and_sugg(
-                cx,
-                REDUNDANT_ALLOCATION,
-                hir_ty.span,
-                "usage of `Rc<Rc<T>>`",
-                "try",
-                snippet_with_applicability(cx, ty.span, "..", &mut applicability).to_string(),
-                applicability,
-            );
-            true
+    let sym_outer = if cx.tcx.is_diagnostic_item(sym::Rc, def_id) {
+        Some(sym::Rc)
+    } else if cx.tcx.is_diagnostic_item(sym::Arc, def_id) {
+        Some(sym::Arc)
+    } else {
+        None
+    };
```

Comment:
```
    let sym_outer = if cx.tcx.is_diagnostic_item(sym::Rc, def_id) {
        Some("Rc")
    } else if cx.tcx.is_diagnostic_item(sym::Arc, def_id) {
        Some("Arc")
    } else {
        None
    };
```
Maybe better?

Reply:
That's definitely better IMO. You can also use an early return to avoid the whole `Option<>` handling. Something similar like this:

```
let sym_outer = if cx.tcx.is_diagnostic_item(sym::Rc, def_id) {
    sym::Rc
} else if cx.tcx.is_diagnostic_item(sym::Arc, def_id) {
    sym::Arc
} else {
    return false;
};
```

That reduces nesting and saves one if statement in line 36

```
if let Some(sym_outer) = sym_outer {
    ...
```

Reply:
Your refactoring looks awesome and is very clean. Nicely done, you can resolve this comment :)